File: solution/14889/main.py
def find_min_diff(start, link, i, S):
    half = len(S) // 2 
    if len(start) == half and len(link) == half:
        diff = 0
        for a in start:
            for b in start:
                diff += S[a][b]
        for a in link:
            for b in link:
                diff -= S[a][b]
        return abs(diff)
    ans = 1 << 30
    if len(start) < half:
        start.append(i)
        ans = min(ans, find_min_diff(start, link, i + 1, S))
        start.pop()
    if len(link) < half:
        link.append(i)
        ans = min(ans, find_min_diff(start, link, i + 1, S))
        link.pop()
    return ans

def main():
    input = open(0, 'rb').readline
    N = int(input())
    S = [[*map(int, input().split())] for _ in range(N)]
    print(find_min_diff([], [], 0, S))

main()

File: solution/14889/main.rs
fn main() {
    use std::io::*;
    let stdin = read_to_string(stdin().lock()).unwrap();
    let mut tokens = stdin.split_whitespace();
    let mut next = || tokens.next().unwrap();
    let n = next().parse::<usize>().unwrap();
    let mut s = vec![];
    for _ in 0..n {
        let mut si = vec![];
        for _ in 0..n {
            let sij = next().parse::<i32>().unwrap();
            si.push(sij);
        }
        s.push(si);
    }
    println!("{}", find_min_diff(&mut vec![], &mut vec![], 0, &s));
}

fn find_min_diff(start: &mut Vec<usize>, link: &mut Vec<usize>, i: usize, S: &[Vec<i32>]) -> i32 {
    let half = S.len() / 2;
    if start.len() == half && link.len() == half {
        let mut diff = 0;
        for &a in start.iter() {
            for &b in start.iter() {
                diff += S[a][b];
            }
        }
        for &a in link.iter() {
            for &b in link.iter() {
                diff -= S[a][b];
            }
        }
        return diff.abs();
    }
    let mut ans = i32::MAX;
    if start.len() < half {
        start.push(i);
        ans = ans.min(find_min_diff(start, link, i + 1, S));
        start.pop();
    }
    if link.len() < half {
        link.push(i);
        ans = ans.min(find_min_diff(start, link, i + 1, S));
        link.pop();
    }
    ans
}
